### ISSUE TITLE
determine a view column's nullability from underlying tables

### ISSUE THREAD

**Issue by elnygren** [hasura/graphql-engine#1965]
Every field of a schema generated from SQL View is nullable/optional even though the underlying table might have NOT NULL fields that are NOT-nullable if the table we're tracked directly.

I've tested and found this problem on the alpha42 version.

This is a problem when working with statically typed languages such as ReasonML. You really don't want to spam option monad's map/flatmap if you don't need to. Basically, this ruins SQL views in a ReasonML / TypeScript (autogenerated types) setup entirely as suddenly everything is nullable.

**Comment by 0x777** [comment]
@elnygren Unfortunately without postgres providing this information, we can't do much at graphql-engine's layer. https://stackoverflow.com/questions/17301323/why-are-my-views-columns-nullable

**Comment by elnygren** [comment]
@0x777 thanks for the answer!

Perhaps there should be some kind of way to produce the GraphQL schema based on 
a) postgresql schema; and
b) user defined settings ?

With b) you could also address other issues such as https://github.com/hasura/graphql-engine/issues/981

Anyhow, I'll leave this into your good hands. :)

**Comment by cbzehner** [comment]
@0x777 in the discussion you link to they explain

> If you tilt your head and squint just right, that kind of makes sense. The column in the view isn't declared NOT NULL. Just the column in the base table.
>The column pg_attribute.attnotnull is updatable. You can set it to TRUE, and that change seems to be reflected in the information_schema views.

While this makes sense, it does appear it should be possible to derive the `null` property of the view columns based on the underlying table. This feels like something that Hasura could provide, even if it's not an immediate priority for y'all.

Could we get this on the long-term roadmap for enhancements or something similar?

**Comment by 0x777** [comment]
@cbzehner 
> While this makes sense, it does appear it should be possible to derive the null property of the view columns based on the underlying table.

Unfortunately this is not easy. Mapping underlying table's columns to columns exposed through a view is only possible in the simplest of the cases and I'm not quite sure if the metadata exposed by postgres about a view is sufficient to determine this.

@elnygren 
> Perhaps there should be some kind of way to produce the GraphQL schema based on
a) postgresql schema; and
b) user defined settings ?

This should definitely be possible.

**Comment by drewwestphal** [comment]
This is also an issue for me. I, too, am using a strongly typed language and this renders views practically unusable. 

I wonder if an override thru the metadata system which implicitly adds not nulls in the query to protect the integrity of the result-set would be a more feasible workaround compared to complex introspection which even postgres opts not to do.

**Comment by drewwestphal** [comment]
Adding in a note from @coco98 in the discord, in case it helps here

![image](https://user-images.githubusercontent.com/229646/58387907-e9da5780-7fcb-11e9-918a-1d1121bffcb0.png)

**Comment by dukedave** [comment]
Ha, I went to all the trouble of making [a repro repo/explanation for this](https://github.com/happymondayco/pg-nullable-view-hasura), and then found this existing issue 🤪

I might have some more insight though:

I see @0x777 already linked this: https://stackoverflow.com/questions/17301323/why-are-my-views-columns-nullable
I also found that, and **a workaround** involving the `pg_attribute` system table. I was able to implement that workaround and confirm that (at least in `psql`) columns in a view **will** show up as `not null`.

I thought that was the solution, but unfortunately Hasura still makes the fields nullable in the the GraphQL schema 😞 

I've made all this available and documented here:
https://github.com/happymondayco/pg-nullable-view-hasura

**Comment by devuxer** [comment]
@coco98, Is this still a priority?

**Comment by sean-bennett112** [comment]
Just wanted to bump this, as it would be incredibly useful to us. We currently separate out our logical API layer and physical table layer in the database, and being able to surface as much type information as possible in views would really improve the client experience.

Note that I'm not expecting that Hasura solve this in the general case by auto-inferring nullability, for instance. Indeed I think that, for us, that would actually be an anti-feature. Rather I'm just looking for a way to specify things like 'this column is not null' in the Hasura metadata layer manually similar to what @drewwestphal described.

**Comment by KATT** [comment]
I just created a view which was like this

```sql
create view "ListingPrice" as
 select id, "priceAmount" as amount, "priceCurrency" as currency from "Listing" where "priceAmount" is not null and "priceCurrency" is not null;
```

I'm pretty sure you could figure out programatically that neither of those fields are nullable 😅

**Comment by TheMrZZ** [comment]
It's indeed a pretty big problem. Views are an inherent part of SQL, and are mandatory to optimize the GraphQL schema. Can this be prioritized ? Even a manual interaction (where we manually select which field is nullable/non-nullable) would be decent enough.

**Comment by willyxiao** [comment]
Yes this would be super helpful to us. Also will really help out with migrations in the DB that can still produce a backwards compatible API via a view.

**Comment by nicuveo** [comment]
Whoops, this was closed accidentally. Sorry about that.

**Comment by clifton** [comment]
this is also causing headaches for me. would be nice if there was a way to force this via metadata as  https://github.com/hasura/graphql-engine/issues/1965#issuecomment-809410886 described

**Comment by jamesknelson** [comment]
For anyone using GraphQL Code Generator, a workaround is to provide an overrides file which specifies that certain columns in the views are non-null. It's not ideal, but it at least makes view usable from within typed languages.

For example, in your graphql config file:

```
schema:
  - "./graphql.codegen.overrides.graphql"
  - "http://localhost:3712/v1/graphql"
```

Then you can specify overrides for views in the new graphql file. For example, if you have a "public_lists" view, and want to specify the `title` and `profile` fields, and `taggings` relationship are non-nullable:

```
type public_lists {
  title: String!
  profile: profiles!
  taggings: [list_taggings!]!
}
```

**Comment by ilijaNL** [comment]
any progress on this to set nullability in hasura metadata?

**Comment by ilijaNL** [comment]
> Just wanted to bump this, as it would be incredibly useful to us. We currently separate out our logical API layer and physical table layer in the database, and being able to surface as much type information as possible in views would really improve the client experience.
> 
> Note that I'm not expecting that Hasura solve this in the general case by auto-inferring nullability, for instance. Indeed I think that, for us, that would actually be an anti-feature. Rather I'm just looking for a way to specify things like 'this column is not null' in the Hasura metadata layer manually similar to what @drewwestphal described.

👍 For this approach

**Comment by SamirTalwar** [comment]
Hey folks, we just announced [Native Queries for PostgreSQL](https://hasura.io/docs/latest/schema/postgres/logical-models/native-queries/) and other databases. They allow you to track an arbitrary SQL query directly in Hasura, giving you complete flexibility but also the usual GraphQL-level selection, sorting, filtering, etc.

Native Queries are based upon [Logical Models](https://hasura.io/docs/latest/schema/postgres/logical-models/index/), which represent the structure of your data. This includes a flag for nullability for each column.

So if you were to create your view as a native query instead, you can specify the nullability explicitly. I hope this helps you solve your problem.

I'll mark this issue as "completed" as explicitly specifying whether a column is nullable should solve your problem, but please feel free to re-open it if you feel you need more information or I've missed something. Cheers!
